File: 02_vehicle/rust/src/main.rs
struct Vehicle {
    brand: String,
    plate: String,
    color: String,
    model: u16,
}

impl Vehicle {
    fn details(&self) -> String {
        format!(
            "Brand: {}\nPlate: {}\nColor: {}\nModel: {}",
            self.brand, self.plate, self.color, self.model
        )
    }
}

fn main() {
    let taxi = Vehicle {
        brand: String::from("Nissan"),
        plate: String::from("RTXB101"),
        color: String::from("Amarillo"),
        model: 2022,
    };
    let moto = Vehicle {
        brand: String::from("Yamanha"),
        plate: String::from("ABC42E"),
        color: String::from("Negra"),
        model: 2021,
    };
    let bus = Vehicle {
        brand: String::from("Chevrolet"),
        plate: String::from("RGD123"),
        color: String::from("De todos los colores"),
        model: 2005,
    };

    println!("\nDetalles taxi");
    println!("{}", taxi.details());

    println!("\nDetalles moto");
    println!("{}", moto.details());

    println!("\nDetalles bus");
    println!("{}", bus.details());
}

File: 02_vehicle/ocaml/bin/main.ml
type vehicle = {
  brand: string;
  plate: string;
  color: string;
  model: int;
}

let details vehicle = 
  match vehicle with
  | { brand; plate; color; model } -> Printf.sprintf "Brand: %s,\nPlate: %s,\nColor: %s,\nModel: %d" brand plate color model

let main () = 
    let taxi = {
        brand = "Nissan";
        plate = "RTXB101";
        color = "Amarillo";
        model = 2022;
    } in

    let moto = {
        brand = "Yamanha";
        plate = "ABC42E";
        color = "Negra";
        model = 2021;
    } in

    let bus = {
        brand = "Chevrolet";
        plate = "RGD123";
        color = "De todos los colores";
        model = 2005;
    } in

    let () = print_endline "\nDetalles taxi" in
    let () = print_endline (details taxi) in

    let () = print_endline "\nDetalles moto" in
    let () = print_endline (details moto) in

    let () = print_endline "\nDetalles bus" in
    let () = print_endline (details bus) in

    ()

let () = main ()
